proposals/thread_parser.py

```
@@ -0,0 +1,188 @@
+            logger.error(f"Error fetching thread {thread_id}: {e}")
+            raise ValueError(f"Error accessing thread: {e}")
+
+    @staticmethod
+    def find_section_header(content: str) -> str:
+        """Find a section header in the content"""
+        for section in ThreadParser.KNOWN_SECTIONS:
+            # Check for variations of the header format
+            patterns = [
+                f"{section}:?\\s*$",  # Plain text with optional colon
+                f"#\\s*{section}:?\\s*$",  # Markdown heading
+                f"\\*\\*{section}\\*\\*:?\\s*$",  # Bold text
+                f"^.*?{section}:?\\s*$",  # More flexible match for Discord's rendering
+            ]
+            for pattern in patterns:
+                if re.search(pattern, content, re.MULTILINE | re.IGNORECASE):
+                    return section
+        return ""
+
+    @staticmethod
+    async def convert_mentions_to_usernames(bot: commands.Bot, content: str) -> str:
+        """Convert Discord mention IDs to usernames"""
+        mention_pattern = r'<@!?(\d+)>'
+        mentions = re.finditer(mention_pattern, content)
+        
+        for mention in mentions:
+            user_id = int(mention.group(1))
+            try:
+                user = await bot.fetch_user(user_id)
+                content = content.replace(mention.group(0), user.name)
+            except:
+                # If we can't fetch the user, leave the mention as is
+                continue
+        
+        return content
+
+    @staticmethod
+    async def parse_thread(bot: commands.Bot, thread_input: str) -> Dict[str, Any]:
+        """Parse a Discord thread to extract messages"""
+        try:
+            thread_id = ThreadParser.extract_thread_id(thread_input)
+            thread = await ThreadParser.fetch_thread(bot, thread_id)
+
+            # Check for multiple proposal type tags
+            governance_tag = False
+            budget_tag = False
+
+            if thread.applied_tags:
```

this should probably throw an explicit error if there are no tags applied (not sure if that's enforced somewhere else), so that we don't get an "index out of range" runtime error when accessing `thread.applied_tags[0]`